keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/proposal/AcceptanceRateTracker.java

```diff
@@ -0,0 +1,34 @@
+package io.improbable.keanu.algorithms.mcmc.proposal;
+
+import com.google.common.collect.Maps;
+import io.improbable.keanu.vertices.Vertex;
+
+import java.util.Map;
+import java.util.Set;
+
+public class AcceptanceRateTracker implements ProposalListener {
+
+    private Map<Set<Vertex>, Integer> numApplied = Maps.newHashMap();
+    private Map<Set<Vertex>, Integer> numRejected = Maps.newHashMap();
+
+    @Override
+    public void onProposalApplied(Proposal proposal) {
+        Set<Vertex> key = proposal.getVerticesWithProposal();
+        Integer previousValue = numApplied.getOrDefault(key, 0);
```

Comment:
If I'm reading this correctly you'll end up doing two Hash code calculations for every Proposal because you're using an Immutable Integer Object to store the numApplied and numRejected.  Given how the Set.hashCode() is defined, this could be pretty expensive?

At the least could we have a little private class instead of Integer to keep the actual count so we can simply call an "increment" style method and avoid the second hash calculation?

Ideally though I'd prefer a different key for this Map - is there any way of refactoring the API here to simplify the key?

Reply:
Thanks, that's a really interesting perspective. I hadn't given any thought to the performance of `Set.hashCode()`!

I checked the javadoc for `Set` and it says: 
> The hash code of a set is defined to be the sum of the hash codes of the elements in the set

That doesn't sound too bad?

Alternatively I could convert it to a `String` key like this, but would that be any better?

    private String generateKey(Set<Vertex> verticesWithProposal) {
        return verticesWithProposal.stream()
            .map(Vertex::getId)
            .sorted()
            .map(Object::toString)
            .collect(Collectors.joining(VERTEX_ID_DELIMITER));
    }

Reply:
To your other point - yes, I can use an `AtomicInteger` for the purpose:

        numApplied.computeIfAbsent(key, s -> new AtomicInteger(0)).incrementAndGet();

Reply:
As discussed I have implemented a custom `Counter` class instead of using `AtomicInteger` - for better performance.